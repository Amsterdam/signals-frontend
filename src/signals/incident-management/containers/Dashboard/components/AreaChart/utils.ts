// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import { Direction } from './ComparisonRate'
import type { ComparisonRateType } from './ComparisonRate'
import type { AreaChartValue as Value, Today } from '../../charts/types'

export const formatData = (values: Value[]): Value[] => {
  return values.map((value, i) =>
    i === values.length - 1
      ? {
          ...value,
          date: `${value.date} 12:00`,
          image: '/assets/images/area-chart-tooltip.svg',
        }
      : { ...value, date: `${value.date} 12:00` }
  )
}

export const getToday = (): Today => {
  const today = new Date()

  return {
    year: today.getFullYear(),
    // Vega lite does not work with zero indexed months
    month: today.getMonth() + 1,
    date: today.getDate(),
    hours: 12,
    minutes: 0o0,
  }
}

export const getPercentage = (values: Value[]): ComparisonRateType => {
  const thisWeek = values.reduce((total, value) => total + value.amount, 0)
  const lastWeek = values.reduce(
    (total, value) => total + value.amount_week_earlier,
    0
  )

  return {
    direction: thisWeek < lastWeek ? Direction.DOWN : Direction.UP,
    percentage: Math.abs(Math.round((thisWeek / lastWeek - 1) * 100)),
  }
}
