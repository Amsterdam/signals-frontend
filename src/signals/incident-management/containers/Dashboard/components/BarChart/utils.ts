// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import configuration from 'shared/services/configuration/configuration'

import type { RawData } from './types'
import { constants } from '../../charts'
import type { BarChartValue } from '../../charts'

export const getQueryList = (queryString: string) => {
  const endpoint = configuration.INCIDENT_PRIVATE_ENDPOINT
  return constants.statusList.map((status) => {
    const params = queryString
      ? `${queryString}&status=${status.query}`
      : `status=${status.query}`
    return `${endpoint}stats/total?${params}`
  })
}

export const formatData = (rawData: RawData[]): BarChartValue[] => {
  return rawData.map((value, i) => {
    return {
      status: constants.statusList[i].label,
      nrOfIncidents: value.total,
    }
  })
}

export const getTotalNrOfIncidents = (rawData: RawData[]): number => {
  return rawData.reduce(
    (aggregatedTotal, value) => aggregatedTotal + value.total,
    0
  )
}
