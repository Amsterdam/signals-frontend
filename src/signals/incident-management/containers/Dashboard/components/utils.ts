// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { IncidentCount } from '../charts/types'

export const getMaxDomain = (values: IncidentCount[]) =>
  Math.max(...values.map((v) => v.amount)) * 1.2
