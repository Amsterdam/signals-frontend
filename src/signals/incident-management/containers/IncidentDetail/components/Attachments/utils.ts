// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { Attachment } from '../../types'

export const getAttachmentFileName = (location: Attachment['location']) =>
  location.split('?')[0].split('/').pop()
