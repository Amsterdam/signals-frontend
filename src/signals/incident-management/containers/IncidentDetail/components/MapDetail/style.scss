.map-detail {
  .map-component {
    #mapdiv {
      height: 474px;
    }
  }
}
