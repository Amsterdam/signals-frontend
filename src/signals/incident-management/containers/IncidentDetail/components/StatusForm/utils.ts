// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2020 - 2021 Gemeente Amsterdam
import type { AlertLevel } from '@amsterdam/asc-ui'

import configuration from 'shared/services/configuration/configuration'
import {
  changeStatusOptionList,
  isStatusClosed,
} from 'signals/incident-management/definitions/statusList'
import { StatusCode } from 'types/status-code'

import * as constants from './constants'

export const emailSentWhenStatusChangedTo = ({
  toStatus,
  fromStatus,
  isSplitIncident,
}: {
  toStatus: StatusCode
  fromStatus: StatusCode
  isSplitIncident: boolean
}): boolean => {
  if (isSplitIncident) return false

  if (
    !configuration.featureFlags.reporterMailHandledNegativeContactEnabled &&
    fromStatus === StatusCode.VerzoekTotHeropenen &&
    toStatus === StatusCode.Afgehandeld
  ) {
    return false
  }

  return Boolean(
    changeStatusOptionList.find(
      ({ email_sent_when_set, key }) => email_sent_when_set && toStatus === key
    )
  )
}

export const textIsRequired = ({
  fromStatus,
  toStatus,
  isSplitIncident,
}: {
  fromStatus: StatusCode
  toStatus: StatusCode
  isSplitIncident: boolean
}): boolean => {
  if (
    isSplitIncident ||
    !configuration.featureFlags.reporterMailHandledNegativeContactEnabled
  ) {
    return [
      StatusCode.Afgehandeld,
      StatusCode.Ingepland,
      StatusCode.Heropend,
      StatusCode.ReactieGevraagd,
    ].includes(toStatus)
  } else {
    return emailSentWhenStatusChangedTo({
      fromStatus,
      toStatus,
      isSplitIncident,
    })
  }
}

type Warning = {
  key: string
  level: AlertLevel
  content: string
  heading?: string
}

export const determineWarnings = ({
  originalStatus,
  toStatus,
  isSplitIncident,
  hasEmail,
  hasOpenChildren,
  categorySlug,
}: {
  originalStatus: StatusCode
  toStatus: StatusCode
  isSplitIncident: boolean
  hasEmail: boolean
  hasOpenChildren: boolean
  categorySlug?: string
}): Warning[] => {
  const warnings: Warning[] = []

  if (isSplitIncident) return []

  if (isStatusClosed(toStatus) && hasOpenChildren) {
    warnings.push({
      key: 'has-open-child-incidents-warning',
      heading: constants.DEELMELDINGEN_STILL_OPEN_HEADING,
      content: constants.DEELMELDINGEN_STILL_OPEN_CONTENT,
      level: 'info',
    })
  }

  if (originalStatus === StatusCode.ReactieGevraagd && hasEmail) {
    warnings.push({
      key: 'has-open-reply-request-warning',
      heading: constants.REPLY_CHANGE_STATUS_HEADING,
      content: constants.REPLY_CHANGE_STATUS_CONTENT,
      level: 'info',
    })
  }

  if (
    configuration.featureFlags.disableClosingCategoryOverigOverig &&
    toStatus === StatusCode.Afgehandeld &&
    categorySlug === 'overig'
  ) {
    warnings.push({
      key: 'has-category-overig',
      heading: constants.CATEGORY_OVERIG_HEADING,
      content: constants.CATEGORY_OVERIG_CONTENT,
      level: 'error',
    })
  } else if (toStatus === StatusCode.Afgehandeld)
    warnings.push({
      key: 'end-status-warning',
      content: constants.AFGEHANDELD_CONTENT,
      level: 'neutral',
    })

  if (toStatus === StatusCode.ReactieGevraagd && !hasEmail) {
    warnings.push({
      key: 'has-no-email-reply-warning',
      heading: constants.REPLY_NO_MAIL_HEADING,
      content: constants.REPLY_NO_MAIL_CONTENT,
      level: 'error',
    })
  }

  return warnings
}

export const getTextConfig = (statusCode: StatusCode) => {
  return statusCode === StatusCode.ReactieGevraagd
    ? {
        label: constants.REPLY_MAIL_LABEL,
        maxLength: constants.REPLY_MAIL_MAX_LENGTH,
        rows: 6,
      }
    : {
        label: constants.DEFAULT_TEXT_LABEL,
        maxLength: constants.DEFAULT_TEXT_MAX_LENGTH,
        rows: 9,
      }
}
