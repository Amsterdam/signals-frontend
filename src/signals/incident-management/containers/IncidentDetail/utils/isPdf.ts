// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam, Vereniging van Nederlandse Gemeenten
export const isPdf = (url: string) => {
  return !!url.match(/[a-zA-Z0-9-_/]+\.pdf/)
}
