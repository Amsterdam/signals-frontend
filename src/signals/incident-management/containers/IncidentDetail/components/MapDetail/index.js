import React from 'react';
import PropTypes from 'prop-types';
import Map from 'components/Map';
import { markerIcon } from 'shared/services/configuration/map-markers';
import { Marker } from '@datapunt/react-maps';

import MAP_OPTIONS from 'shared/services/configuration/map-options';
import { locationType } from 'shared/types';
import './style.scss';

const MapDetail = ({ value, className, zoom, icon, hasZoomControls }) => {
  const location = value?.geometrie?.coordinates;
  const lat = location && location[1];
  const lng = location && location[0];
  const options = {
    ...MAP_OPTIONS,
    zoom,
    attributionControl: false,
    center: [lat, lng],
  };

  return lat && lng ? (
    <Map mapOptions={options} canBeDragged={false} className={className} hasZoomControls={hasZoomControls}>
      <Marker args={[{ lat, lng }]} options={{ icon }} />
    </Map>
  ) : null;
};

MapDetail.defaultProps = {
  className: '',
  hasZoomControls: false,
  icon: markerIcon,
};

MapDetail.propTypes = {
  className: PropTypes.string,
  hasZoomControls: PropTypes.bool,
  icon: PropTypes.shape({}), // leaflet icon object
  value: locationType.isRequired,
  zoom: PropTypes.number.isRequired,
};

export default MapDetail;
