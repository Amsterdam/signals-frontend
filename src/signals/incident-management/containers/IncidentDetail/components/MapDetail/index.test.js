import React from 'react';
import { render } from '@testing-library/react';
import { withAppContext } from 'test/utils';

import { markerIcon } from 'shared/services/configuration/map-markers';
import MapDetail from './index';

describe('<MapDetail />', () => {
  const props = {
    value: {
      geometrie: { coordinates: [4, 42] },
    },
    zoom: 15,
    icon: markerIcon,
  };

  beforeEach(() => {
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    const { container, getByTestId } = render(withAppContext(<MapDetail {...props} />));

    // Map
    expect(getByTestId('map-base')).toBeInTheDocument();

    // Marker
    expect(container.querySelector('.map-marker-select')).toBeInTheDocument();
  });

  it('should not render without value', () => {
    props.value = {};
    const { container, queryByTestId } = render(withAppContext(<MapDetail {...props} />));

    // Map
    expect(queryByTestId('map')).not.toBeInTheDocument();

    // Marker
    expect(container.querySelector('.map-marker-select')).not.toBeInTheDocument();
  });
});
