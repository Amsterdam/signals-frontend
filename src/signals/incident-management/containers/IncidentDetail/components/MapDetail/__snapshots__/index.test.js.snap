// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MapDetail /> should render correctly 1`] = `
<div
  className="map-detail"
>
  <Map
    hideAttribution={false}
    hideZoomControls={false}
    latlng={
      Object {
        "latitude": 0,
        "longitude": 0,
      }
    }
    useSmallMarker={false}
    zoom="16"
  />
</div>
`;

exports[`<MapDetail /> should render correctly without location 1`] = `
<div
  className="map-detail"
/>
`;
